### ISSUE TITLE
AKSPublicIPPrefixSpec test spec is buggy

### ISSUE THREAD

**Issue by jackfrancis** [kubernetes-sigs/cluster-api-provider-azure#3047]
/kind bug

[Before submitting an issue, have you checked the [Troubleshooting Guide](https://capz.sigs.k8s.io/topics/troubleshooting.html)?]

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

@nojnhuh has already done a bit of high fidelity investigation, copied from public slack conversation:

> This is the test where we create a node pool with more nodes than available IPs in the prefix, and expect that to fail. I do see the “Ready” condition going to “Failed” like the test expects, but then the operation is immediately retried so it goes back to “Creating” and the polling in the test doesn’t happen to catch it when it’s “Failing.” In the test, it looks like we’re assuming it to fail once and not be retried. The failing runs show the expected error coming up multiple times in the controller-manager logs, but only once in a previous passing run, so that seems to agree with that assumption.
AFAICT, the error messages between the passing and failing runs are the same and being surfaced at the same point, so I’m still looking into why they’re being treated differently.
The passing run I was looking at: [https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluste[…]890/pull-cluster-api-provider-azure-e2e-exp/1613978841884135424](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/2890/pull-cluster-api-provider-azure-e2e-exp/1613978841884135424)
Failing: [https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluste[…]038/pull-cluster-api-provider-azure-e2e-exp/1614059198734143488](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/3038/pull-cluster-api-provider-azure-e2e-exp/1614059198734143488)

> It seems like both the passing and failing cases follow the path leading to [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/8e4ea53af9477ae79a2384d16ad561c8d5faeab8/exp/controllers/azuremanagedmachinepool_controller.go#L285), and since returning and error in Reconcile triggers a requeue, I’m not sure why we weren’t seeing this being retried before.

> And the diff between the last passing and first failing runs looks like a big nothing burger: https://github.com/kubernetes-sigs/cluster-api-provider-azure/compare/f2a45077f...8e4ea53af

**What did you expect to happen:**

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

**Comment by nojnhuh** [comment]
`git bisect` speaks. This one-line change is causing this: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3030/files#diff-2c4585415b3fabfd1913f95eb20bc57daac00c2facd3ffa45d3bb00bb1d41b63

Before that change, the status condition would change to "Failed" and show the relevant IP prefix error, then quickly converge to the following, still "Failed": 
```
failed to find vm scale set in resource group MC_aks-15925_aks-15925_eastus matching pool named pool2. Object will be requeued after 20s
```

The above change is forcing unnecessary re-reconciliations when no taints are specified because of a nil/empty pointer-to-slice discrepancy:
```diff
  containerservice.AgentPool{
  	Response: {},
  	ManagedClusterAgentPoolProfileProperties: &containerservice.ManagedClusterAgentPoolProfileProperties{
  		... // 28 identical fields
  		Tags:                      nil,
  		NodeLabels:                nil,
- 		NodeTaints:                &nil,
+ 		NodeTaints:                nil,
  		ProximityPlacementGroupID: nil,
  		KubeletConfig:             nil,
  		... // 6 identical fields
  	},
  	ID:   nil,
  	Name: nil,
  	Type: nil,
  }
```

So in the next reconcile after the IP prefix error, that change was prompting a new PUT because of the erroneous diff, where before there would be no diff so it continues all the way to checking that the backing VMSS exists and fails there each time thereafter.

**Comment by jackfrancis** [comment]
@nojnhuh nice find! so to me the larger issue is that our particular E2E spec data structures need to keep pace w/ the changes we make in the CAPZ<--->AKS API plumbing.

Is there a simpler change to the E2E spec for the add'l pool we can make in https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3048 so that we don't have to entirely disable the scenario?

**Comment by nojnhuh** [comment]
This was just a bug in the controller AFAICT, nothing in the test was ill-defined or regressed. I didn't make it clear, but that diff above is from an ad hoc print statement in the controller here:
https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/6a759e133159035e3aa772470ebb25249a27f986/azure/services/agentpools/spec.go#L235
